services/motion/builtin/state/state.go

```diff
@@ -0,0 +1,619 @@
+// If the ExecutionID is provided, returns the plans of the ExecutionID rather
+// than the most recent execution
+// If LastPlanOnly is provided then only the last plan is returned for the execution
+// with the ExecutionID if it is provided, or the last execution
+// for that component otherwise.
+func (s *State) PlanHistory(req motion.PlanHistoryReq) ([]motion.PlanWithStatus, error) {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+	cs, exists := s.componentStateByComponent[req.ComponentName]
+	if !exists {
+		return nil, ErrUnknownResource
+	}
+
+	executionID := req.ExecutionID
+
+	// last plan only
+	if req.LastPlanOnly {
+		if ex := cs.lastExecution(); executionID == uuid.Nil || executionID == ex.id {
+			return []motion.PlanWithStatus{ex.history[0]}, nil
+		}
+
+		// if executionID is provided & doesn't match the last execution for the component
+		if ex, exists := cs.executionsByID[executionID]; exists {
+			return []motion.PlanWithStatus{ex.history[0]}, nil
+		}
+		return nil, ErrNotFound
+	}
+
+	// specific execution id when lastPlanOnly is NOT enabled
+	if executionID != uuid.Nil {
+		if ex, exists := cs.executionsByID[executionID]; exists {
+			history := make([]motion.PlanWithStatus, len(ex.history))
+			copy(history, ex.history)
+			return history, nil
+		}
+		return nil, ErrNotFound
+	}
+
+	ex := cs.lastExecution()
+	history := make([]motion.PlanWithStatus, len(cs.lastExecution().history))
+	copy(history, ex.history)
+	return history, nil
+}
+
+// ListPlanStatuses returns the status of plans created by MoveOnGlobe requests
+// that are executing OR are part of an execution which changed it state
+// within the a 24HR TTL OR until the robot reinitializes.
+// If OnlyActivePlans is provided, only returns plans which are in non terminal states.
+func (s *State) ListPlanStatuses(req motion.ListPlanStatusesReq) ([]motion.PlanStatusWithID, error) {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	statuses := []motion.PlanStatusWithID{}
+	if req.OnlyActivePlans {
+		for name := range s.componentStateByComponent {
+			if e, err := s.activeExecution(name); err == nil {
+				statuses = append(statuses, motion.PlanStatusWithID{
+					ExecutionID:   e.id,
+					ComponentName: e.componentName,
+					PlanID:        e.history[0].Plan.ID,
+					Status:        e.history[0].StatusHistory[0],
+				})
+			}
+		}
+		return statuses, nil
+	}
+
+	for _, cs := range s.componentStateByComponent {
+		for _, executionID := range cs.executionIDHistory {
+			e, exists := cs.executionsByID[executionID]
+			if !exists {
+				return nil, errors.New("state is corrupted")
+			}
+			for _, pws := range e.history {
+				statuses = append(statuses, motion.PlanStatusWithID{
+					ExecutionID:   e.id,
+					ComponentName: e.componentName,
+					PlanID:        pws.Plan.ID,
+					Status:        pws.StatusHistory[0],
+				})
+			}
+		}
+	}
+
+	return statuses, nil
+}
+
+// validateNoActiveExecutionID returns an error if there is already an active
```

Comment:
To be sure I understand, limitations on this would be:
- if the resource is being moved by another motion service, this will return nil
- if the resource is being moved by calls directly to that resource, this will return nil
- if the resource's frame system parent is being moved by *this* motion service, and that plan is therefore taking into account where this resources' geometries are, this will still return nil.

Reply:
1. True: A State is scoped to a given motion service so it has no knowledge of what another motoin service instance is doing.
2. True: If the move wasn't created via MoveOnGlobe then the State won't know about it & this function will return nil
3. True: I wasn't aware that was a requirement

Reply:
I don't think (3) is a requirement, as rover-on-a-rover would be a pretty niche case, and arm-on-a-rover is something we've not gone out of our way to support yet.